docs/design-proposals/apiserver/openstack-api-support.md

```diff
@@ -0,0 +1,385 @@
+---
+title: Support Openstack Compute APIs in Arktos
+authors:
+  - "@yb01"
+---
+
+# Support Openstack Compute APIs in Arktos
+## Background and context
+To boost Arktos VM adoption and to let Openstack users to easily explore Arktos VM support without a lot of changes from the users’ applications, it is decided to add a new set of APIs in Arktos to match Openstack Server related operations. Besides better user experiences, this new APIs will also help existing Openstack perf tests to be relatively easily converted to run against Arktos.
+
+## Goal and non-goals
+
+Goals:
+
+1. New set of REST interfaces in Arktos to support Openstack requests for Server CRUD
+1. New set of REST interfaces in Arktos to support Openstack requests for Server Actions
+1. Nova or Nova test tool can switch to Arktos for basic VM operation, by switching to the Arktos URL
+
+Non-goals:
+
+1. Openstack client-side SDK or other dev tools are not in the scope of this design doc.
+1. Openstack API other than VM CRUD, VM Actions, is not in the scope of this design doc.
+
+## New APIs in Arktos
+### New API route path:
+The new route for servers
+
+* 	{Arktos-service-url}/servers
+* 	{Arktos-service-url}/servers/{server-id}/detail
+
+The new route for server actions
+
+* {Arktos-service-url}/servers/{server-id}/action
+
+
+### API call chain
+The API call chains of /servers will be the same as the other API requests, which is created in the DefaultBuildHandlerChain() function
+
+
+### VM list, create, delete
+
+|Methods |URL.     |Function                     |Supported in Arktos|
+|--------|---------|-----------------------------|-------------------|
+|GET	  |/servers  |	List VMs for all tenants  | No |
+|GET	  |/servers  |	List all VMs for a tenant/namespace | yes |
+|POST	  |/servers	 |Create a VM	                |yes |
+|POST	  |/servers	 |Create multiple VMs.        |no
+|GET	  |/servers/detail	| List VM details for a tenant/namespace | Not for this release|
+|GET	  |/servers/{server-id}	|Show VM details |Yes|
+|PUT	  |/servers/{server-id} |	Update a VM | Not in this release |
+|DELETE   |/servers  |/{server-id}	Delete a VM |Yes|
+|GET	  |/servers	 |List VM for all tenants |No in this release|
+
+### VM actions
+|Methods |URL.     |Function                     |Supported in Arktos|
+|--------|---------|-----------------------------|-------------------|
+|POST	 |/servers/{server-id}/action	|Perform specific action on a given VM |	Yes (subset of actions) * |
+
+*Currently only “reboot, stop, start, snapshot, restore”
+
+### Errors
+Will relay whatever errors from Arktos, which is standard http errors
```

Comment:
Nova has its own error codes, mostly common used, e.g., 

* Normal response codes: 200
* Error response codes: badRequest(400), unauthorized(401), forbidden(403)

We may need to follow that so that we could reuse its SDK. 

Example: https://docs.openstack.org/api-ref/compute/?expanded=list-servers-detailed-detail#show-details-of-specific-api-version

Reply:
Those 200 or 400 error code are standard and used in Arktos as well. i will add a note, to support some Openstack specific error code ( this would need a error converting layer ) just like we did for the request and response.

Reply:
on going effort. if there turns out many openstack specific errors used in the test tools. we can consider add them into the specific error handler code.